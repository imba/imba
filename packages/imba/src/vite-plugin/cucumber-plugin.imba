import { toRollupError } from "./utils/error";
import {
	preResolveOptions,
	resolveOptions,
} from "./utils/options";
import { createCompileImba } from "./utils/compile";
import Handlebars from "handlebars";
import rawTestTemplate from './cucumberTemplate.txt?raw'
import {parse} from 'gherkin-io'
import Gherkin from '@cucumber/gherkin'

def convertToEnglishKeyword(keyword, language = 'en')
	if language == 'en'
		return keyword.toLowerCase!.trim!
	const dialect = Gherkin.dialects[language]
	for own key, values of dialect
		if values..map(do $1.trim!).includes(keyword)
			return key.toLowerCase!.trim!
	null

Handlebars.registerHelper('decoratedSuite') do(keyword, item)
	if item.tags..find(do $1.name == 'only')
		keyword += ".only"
	
	if item.tags..find(do $1.name == 'fails')
		keyword += ".fails"

	if item.tags..find(do $1.name == 'todo')
		keyword += ".todo"

	if item.tags..find(do $1.name == 'todo')
		keyword += ".todo"

	keyword

Handlebars.registerHelper('escape') do $1.replaceAll("'", "\\'")

Handlebars.registerHelper('isScenario') do(a, lang) 
	convertToEnglishKeyword(a, lang) == 'scenario' or convertToEnglishKeyword(a, lang) == 'example'

Handlebars.registerHelper('isScenarioOutline') do(a, lang)
	convertToEnglishKeyword(a, lang) == 'scenario outline' or convertToEnglishKeyword(a, lang) == 'scenario template'

Handlebars.registerHelper('getOriginalValue') do(val)
	if isNaN(val)
		`'"{val}"'`
	else val 
Handlebars.registerHelper('getDescriptionFromCells') do(cells)
	cells.map(do $1.value).join(', ')

Handlebars.registerHelper('getCellValue') do(header, index)
	header.cells[index].value

const L = console.log

const testTemplate = Handlebars.compile(rawTestTemplate)

export def generateImbaCode(id, content)
	let doc = await parse(content, id)
	const feature = doc.feature
	const lang = feature.language
	const stepDefsGlob = './step_definitions/**/*.imba'
	const backgroundEl = feature.elements.filter(do convertToEnglishKeyword($1.keyword, lang) == 'background')[0]
	const baseContextPath = './context.imba'

	def hasDecoractor(item, val)
		val = [val] unless Array.isArray(val)

		return item.tags..find(do val.indexOf($1.name) != -1)
		for el in item.elements or []
			return hasDecoractor(el)
		no

	const code = testTemplate({
		feature
		backgroundEl
		stepDefsGlob
		baseContextPath
		retry: hasDecoractor(feature, ['retry'])..value
		hasNoTimeout: !!hasDecoractor(feature, ['notimeout', 'debug', 'noTimeout'])
	})

def _transform(id, content, compileImba, options)
	const code = await generateImbaCode(id, content)
	let compiledData
	const imbaRequest = {
		id: id
		filename: id
		normalizedFilename: id
		query: {}
	}
	try compiledData = await compileImba(imbaRequest, code, options) catch e
		throw toRollupError(e, options);
	return compiledData.compiled.js

export def parseFeatureIntoSteps(feature)
	let id = "dynamic.feature"
	id = feature.file if feature.file
	feature = feature.content if feature.content
	const lang = feature.language
	const doc = await parse(feature, id)	
	const bgElements = doc.feature.elements
		.filter(do convertToEnglishKeyword($1.keyword, lang) == 'background')
		.flatMap do $1.steps

	doc.feature.elements.flatMap do(element)
		let k = convertToEnglishKeyword(element.keyword, lang)
		if k == 'scenario'
			bgElements.concat element.steps
		elif k == 'background' 
			[]
		elif k == 'scenario outline' or k == 'scenario template'
			element.examples.flatMap do(example)
				example.body.flatMap do({cells})
					bgElements.concat element.steps.map do(step)
						let text = step.text
						for {value}, i in cells
							const v = if isNaN(value) then `'"{value}"'` else value 
							text = text.replaceAll("<{example.header.cells[i].value}>", v)
						{text, keyword:step.keyword}

export default def cucumberPlugin(inlineOptions = {})
	let options
	let compileImba\( (imbaRequest: ImbaRequest,code: string,options: Partial<ResolvedOptions>,) => Promise<CompileData>)

	
	def config(config, configEnv)
		options = await preResolveOptions(inlineOptions, config, configEnv)

	def configResolved(config)
		options = resolveOptions(options, config);
		compileImba = createCompileImba(options);
		
	def transform(content, id)
		if id.endsWith('.feature')
			await _transform(id, content, compileImba, options)

	return 
		name: "vite-plugin-imba-cucumber"
		enforce: 'pre'
		config: config
		configResolved: configResolved
		transform: transform
