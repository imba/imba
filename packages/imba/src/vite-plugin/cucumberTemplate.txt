import {parseFeatureIntoSteps} from "imba/plugin"

const modules = import.meta.glob('{{stepDefsGlob}}', {eager: yes})
const baseContexts = import.meta.glob('./**/*.context.imba', {eager: yes})
let s
let c
let elName
let contexts = {}
let text
let stepRes
let keyword
let currentContext

const decorators = {}

{{#if retry}}
vi.setConfig
	retry: {{retry}}
{{/if}}
{{#if hasNoTimeout}}
vi.setConfig 
	testTimeout: 50000000
	hookTimeout: 50000000
{{/if}}

def executeStep(text, keyword, context)
	let s = global.Steps.find(text, keyword).stepDefinition
	if const ext = s.target.extendContext
		await ext(context)
	let res = await s.target[s.fname]
		.apply(
			context,
			# variables
			[s.cucumberExpression.match(text).map(do $1.getValue!)]
		)

	# Assertions throw an error if accessed (they use a proxy)
	return if res..__flags
	return unless const features = res..__features
	await features.reduce(&, do 1) do(acc, feature)
		await acc
		const steps = await parseFeatureIntoSteps(feature)
		await steps.reduce(&, do 1) do(acc, step)
			await acc
			await executeStep(step.text, step.keyword, context)


beforeAll do(suite)
	{{#each feature.tags}}
	decorators['{{this.name}}'] = '{{this.value}}' or yes
	   
	{{/each}}
beforeEach do(suite)

	let contextName = 'BaseContext'
	const baseContext = Object.values(baseContexts).find(do $1[contextName])..[contextName]
	{{#each feature.elements}}
	elName = '{{{ this.keyword }}}: {{{ this.name }}}'
	if suite.meta.name == elName or suite.meta.suite..suite..name == elName
		suite.meta.decorators = decorators
		{{#each this.tags}}
		suite.meta.decorators['{{this.name}}'] = '{{this.value}}' or yes
		
		{{/each}}
		{{!-- L "creating context c{{ @index }}" --}}
		contexts.c{{ @index }} = {
			context: global.__ccContext(baseContext, suite.meta)
			name: elName
		}
	{{/each}}

	{{#if backgroundEl}}
	   {{#each backgroundEl.steps}}
	# find the context for the current test this beforeEach hook is running before
	c = Object.keys(contexts).find(do 
		contexts[$1].name == suite.meta.name or contexts[$1].name == suite.meta.suite..suite..name 
	)
	await executeStep('{{{escape this.text}}}', '{{this.keyword}}', contexts[c].context)

	   {{/each}}
	{{/if}}

{{decoratedSuite 'describe' this}} 'Feature: {{ feature.name }}' do
{{#each feature.elements}}
	{{#if (isScenarioOutline this.keyword ../feature.language)}}
	{{decoratedSuite 'describe' this}} '{{ this.keyword }}: {{ escape this.name }}' do
		{{#each this.examples}}
		{{decoratedSuite 'describe' this}} 'Examples group {{ @index }}: {{ escape this.name}}' do
			{{#each this.body}}
			{{decoratedSuite 'test' this}} 'Example {{ @index }}: {{getDescriptionFromCells this.cells}}' do(suite)
				{{#each ../../this.steps}}
				text = '{{{ escape this.text }}}'
					{{#each ../this.cells}}
				text = text.replaceAll('<{{getCellValue ../../../this.header @index}}>', {{{getOriginalValue this.value}}})
					{{/each}}
				keyword = '{{{ this.keyword }}}'
				currentContext = contexts.c{{ @../../../index }}.context
				await executeStep(text, keyword, currentContext)
				{{/each}}
			{{/each}}
		{{/each}}
	{{/if}}
	{{#if (isScenario this.keyword ../feature.language)}}
	{{decoratedSuite 'test' this}} '{{ this.keyword }}: {{ escape this.name }}' do
		{{#each this.steps}}
		text = '{{{ escape this.text }}}'
		keyword = '{{{ this.keyword }}}'
		currentContext = contexts.c{{ @../index }}.context
		await executeStep(text, keyword, currentContext)
		{{/each}}
	{{/if}}
{{/each}}