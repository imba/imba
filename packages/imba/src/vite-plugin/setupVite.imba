import np from 'path'
import * as Vite from 'vite'
import url from 'node:url'
import nfs from 'node:fs'
import getport from 'get-port'
import { __served__ } from '../imba/utils'
import {getConfigFilePath} from '../utils/vite'
let vite

export def setupVite(srv, options, cb)
	global.__vite__ = yes
	if typeof options == 'string'
		options = {mode: options}
	elif typeof options == 'function'
		options = {mode: "development"}
	const prod? = options.mode == "production" or options.prod == yes
	const mode = prod? ? "production" : "development"

	unless vite
		const port = await getport(port: getport.makeRange(24000, 26000))

		const serverConfig = await getConfigFilePath("server", {command: "server", mode})

		let serverOptions = options.serverOptions or { server: { hmr: {port} } }

		let clientConfig = await getConfigFilePath "client", {mode, vite:yes}

		const vite-options = {
			...clientConfig
			...serverOptions,
			appType: "custom"
			server: {
				...serverConfig.server,
				...serverOptions.server,
				middlewareMode: true
			}
		}

		vite = await Vite.createServer vite-options

	let clientDist = np.join(vite.config.root, options.outDir or 'dist', 'public')

	if prod?
		vite.close().then do vite = null
		global.__vite_manifest__ ||= JSON.parse nfs.readFileSync(np.join(clientDist, 'manifest.json'), 'utf-8')
		global.IMBA_PUBDIR = clientDist
		cb(clientDist) if cb
		return clientDist
	else
		srv.use do(req, res, next)
			__served__.add req.url
			next()
		srv.use vite.middlewares
		srv.use do(req, res, next)
			__served__.delete req.url
			next()
	return clientDist
