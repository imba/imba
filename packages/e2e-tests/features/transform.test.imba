import {generateImbaCode} from '../../imba/src/vite-plugin/cucumber-plugin'
import np from 'node:path'
import nfs from 'node:fs'

def transform(path)
	path = np.resolve np.join('features', path)
	const content = await nfs.promises.readFile(path, 'utf-8')
	await generateImbaCode(path, content)

# It makes sense to write these tests manually
# if this is broken then feature tests won't run
describe "Cucumber plugin " do
	test "Add.feature" do
		const code = await transform("add.feature")
		expect(code).toMatchSnapshot!

	test "Tags.feature" do
		const code = await transform("tags.feature")
		expect(code).toMatchSnapshot!

	test "scenario-outline feature" do
		const code = await transform("scenario-outline.feature")
		expect(code).toMatchSnapshot!

	test "support other languages" do
		const code = await transform("ar.feature")
		expect(code).toMatchSnapshot!
