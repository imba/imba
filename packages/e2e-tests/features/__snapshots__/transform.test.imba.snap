// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Cucumber plugin  > Add.feature 1`] = `
"import {parseFeatureIntoSteps} from \\"imba/plugin\\"

const modules = import.meta.glob('./step_definitions/**/*.imba', {eager: yes})
const baseContexts = import.meta.glob('./**/*.context.imba', {eager: yes})
let s
let c
let elName
let contexts = {}
let text
let stepRes
let keyword
let currentContext

const decorators = {}


def executeStep(text, keyword, context)
	let s = global.Steps.find(text, keyword).stepDefinition
	if const ext = s.target.extendContext
		await ext(context)
	let res = await s.target[s.fname]
		.apply(
			context,
			# variables
			[s.cucumberExpression.match(text).map(do $1.getValue!)]
		)

	# Assertions throw an error if accessed (they use a proxy)
	return if res..__flags
	return unless const features = res..__features
	await features.reduce(&, do 1) do(acc, feature)
		await acc
		const steps = await parseFeatureIntoSteps(feature)
		await steps.reduce(&, do 1) do(acc, step)
			await acc
			await executeStep(step.text, step.keyword, context)


beforeAll do(suite)
beforeEach do(suite)

	let contextName = 'BaseContext'
	const baseContext = Object.values(baseContexts).find(do $1[contextName])..[contextName]
	elName = 'Background: Sets initial value for the sum'
	if suite.meta.name == elName or suite.meta.suite..suite..name == elName
		suite.meta.decorators = decorators
		contexts.c0 = {
			context: global.__ccContext(baseContext, suite.meta)
			name: elName
		}
	elName = 'Example: Adding two numbers works with background'
	if suite.meta.name == elName or suite.meta.suite..suite..name == elName
		suite.meta.decorators = decorators
		contexts.c1 = {
			context: global.__ccContext(baseContext, suite.meta)
			name: elName
		}
	elName = 'Scenario: Supports multiple scenarios'
	if suite.meta.name == elName or suite.meta.suite..suite..name == elName
		suite.meta.decorators = decorators
		contexts.c2 = {
			context: global.__ccContext(baseContext, suite.meta)
			name: elName
		}

	# find the context for the current test this beforeEach hook is running before
	c = Object.keys(contexts).find(do 
		contexts[$1].name == suite.meta.name or contexts[$1].name == suite.meta.suite..suite..name 
	)
	await executeStep('the initial value is 10', 'Given', contexts[c].context)


describe 'Feature: Addition' do
	test 'Example: Adding two numbers works with background' do
		text = 'I add 12 and 4'
		keyword = 'Given'
		currentContext = contexts.c1.context
		await executeStep(text, keyword, currentContext)
		text = 'I should obtain 26'
		keyword = 'Then'
		currentContext = contexts.c1.context
		await executeStep(text, keyword, currentContext)
	test 'Scenario: Supports multiple scenarios' do
		text = 'I add 12 and 10'
		keyword = 'Given'
		currentContext = contexts.c2.context
		await executeStep(text, keyword, currentContext)
		text = 'I should obtain 32'
		keyword = 'Then'
		currentContext = contexts.c2.context
		await executeStep(text, keyword, currentContext)
"
`;

exports[`Cucumber plugin  > Tags.feature 1`] = `
"import {parseFeatureIntoSteps} from \\"imba/plugin\\"

const modules = import.meta.glob('./step_definitions/**/*.imba', {eager: yes})
const baseContexts = import.meta.glob('./**/*.context.imba', {eager: yes})
let s
let c
let elName
let contexts = {}
let text
let stepRes
let keyword
let currentContext

const decorators = {}

vi.setConfig
	retry: 4
vi.setConfig 
	testTimeout: 50000000
	hookTimeout: 50000000

def executeStep(text, keyword, context)
	let s = global.Steps.find(text, keyword).stepDefinition
	if const ext = s.target.extendContext
		await ext(context)
	let res = await s.target[s.fname]
		.apply(
			context,
			# variables
			[s.cucumberExpression.match(text).map(do $1.getValue!)]
		)

	# Assertions throw an error if accessed (they use a proxy)
	return if res..__flags
	return unless const features = res..__features
	await features.reduce(&, do 1) do(acc, feature)
		await acc
		const steps = await parseFeatureIntoSteps(feature)
		await steps.reduce(&, do 1) do(acc, step)
			await acc
			await executeStep(step.text, step.keyword, context)


beforeAll do(suite)
	decorators['debug'] = '' or yes
	   
	decorators['makeVersion10'] = '' or yes
	   
	decorators['retry'] = '4' or yes
	   
beforeEach do(suite)

	let contextName = 'BaseContext'
	const baseContext = Object.values(baseContexts).find(do $1[contextName])..[contextName]
	elName = 'Scenario: Accessing a value in context works'
	if suite.meta.name == elName or suite.meta.suite..suite..name == elName
		suite.meta.decorators = decorators
		contexts.c0 = {
			context: global.__ccContext(baseContext, suite.meta)
			name: elName
		}


describe 'Feature: Context' do
	test 'Scenario: Accessing a value in context works' do
		text = 'The \\"version\\" in context is \\"10\\"'
		keyword = 'Then'
		currentContext = contexts.c0.context
		await executeStep(text, keyword, currentContext)
		text = 'The \\"retry\\" in context is \\"4\\"'
		keyword = 'And'
		currentContext = contexts.c0.context
		await executeStep(text, keyword, currentContext)
"
`;

exports[`Cucumber plugin  > scenario-outline feature 1`] = `
"import {parseFeatureIntoSteps} from \\"imba/plugin\\"

const modules = import.meta.glob('./step_definitions/**/*.imba', {eager: yes})
const baseContexts = import.meta.glob('./**/*.context.imba', {eager: yes})
let s
let c
let elName
let contexts = {}
let text
let stepRes
let keyword
let currentContext

const decorators = {}


def executeStep(text, keyword, context)
	let s = global.Steps.find(text, keyword).stepDefinition
	if const ext = s.target.extendContext
		await ext(context)
	let res = await s.target[s.fname]
		.apply(
			context,
			# variables
			[s.cucumberExpression.match(text).map(do $1.getValue!)]
		)

	# Assertions throw an error if accessed (they use a proxy)
	return if res..__flags
	return unless const features = res..__features
	await features.reduce(&, do 1) do(acc, feature)
		await acc
		const steps = await parseFeatureIntoSteps(feature)
		await steps.reduce(&, do 1) do(acc, step)
			await acc
			await executeStep(step.text, step.keyword, context)


beforeAll do(suite)
beforeEach do(suite)

	let contextName = 'BaseContext'
	const baseContext = Object.values(baseContexts).find(do $1[contextName])..[contextName]
	elName = 'Background: '
	if suite.meta.name == elName or suite.meta.suite..suite..name == elName
		suite.meta.decorators = decorators
		contexts.c0 = {
			context: global.__ccContext(baseContext, suite.meta)
			name: elName
		}
	elName = 'Scenario: Concat weird strings'
	if suite.meta.name == elName or suite.meta.suite..suite..name == elName
		suite.meta.decorators = decorators
		contexts.c1 = {
			context: global.__ccContext(baseContext, suite.meta)
			name: elName
		}
	elName = 'Scenario Outline: Adding two numbers works on many cases'
	if suite.meta.name == elName or suite.meta.suite..suite..name == elName
		suite.meta.decorators = decorators
		contexts.c2 = {
			context: global.__ccContext(baseContext, suite.meta)
			name: elName
		}
	elName = 'Scenario Outline: Adding two numbers works on many cases'
	if suite.meta.name == elName or suite.meta.suite..suite..name == elName
		suite.meta.decorators = decorators
		contexts.c3 = {
			context: global.__ccContext(baseContext, suite.meta)
			name: elName
		}

	# find the context for the current test this beforeEach hook is running before
	c = Object.keys(contexts).find(do 
		contexts[$1].name == suite.meta.name or contexts[$1].name == suite.meta.suite..suite..name 
	)
	await executeStep('the initial value is 0', 'Given', contexts[c].context)


describe 'Feature: Scenario outline' do
	test 'Scenario: Concat weird strings' do
		text = 'I concatenate \\"(scrim-toggler[sref=\\\\'browserbutton\\\\']\\" and \\")\\"'
		keyword = 'Given'
		currentContext = contexts.c1.context
		await executeStep(text, keyword, currentContext)
		text = 'The \\"str\\" in context is \\"(scrim-toggler[sref=\\\\'browserbutton\\\\'])\\"'
		keyword = 'Then'
		currentContext = contexts.c1.context
		await executeStep(text, keyword, currentContext)
	describe 'Scenario Outline: Adding two numbers works on many cases' do
		describe 'Examples group 0: Basic examples' do
			test 'Example 0: 1, 2, 3' do(suite)
				text = 'I add <a> and <b>'
				text = text.replaceAll('<a>', 1)
				text = text.replaceAll('<b>', 2)
				text = text.replaceAll('<result>', 3)
				keyword = 'Given'
				currentContext = contexts.c2.context
				await executeStep(text, keyword, currentContext)
				text = 'I should obtain <result>'
				text = text.replaceAll('<a>', 1)
				text = text.replaceAll('<b>', 2)
				text = text.replaceAll('<result>', 3)
				keyword = 'Then'
				currentContext = contexts.c2.context
				await executeStep(text, keyword, currentContext)
			test 'Example 1: 12, 2, 14' do(suite)
				text = 'I add <a> and <b>'
				text = text.replaceAll('<a>', 12)
				text = text.replaceAll('<b>', 2)
				text = text.replaceAll('<result>', 14)
				keyword = 'Given'
				currentContext = contexts.c2.context
				await executeStep(text, keyword, currentContext)
				text = 'I should obtain <result>'
				text = text.replaceAll('<a>', 12)
				text = text.replaceAll('<b>', 2)
				text = text.replaceAll('<result>', 14)
				keyword = 'Then'
				currentContext = contexts.c2.context
				await executeStep(text, keyword, currentContext)
			test 'Example 2: 13, 1, 14' do(suite)
				text = 'I add <a> and <b>'
				text = text.replaceAll('<a>', 13)
				text = text.replaceAll('<b>', 1)
				text = text.replaceAll('<result>', 14)
				keyword = 'Given'
				currentContext = contexts.c2.context
				await executeStep(text, keyword, currentContext)
				text = 'I should obtain <result>'
				text = text.replaceAll('<a>', 13)
				text = text.replaceAll('<b>', 1)
				text = text.replaceAll('<result>', 14)
				keyword = 'Then'
				currentContext = contexts.c2.context
				await executeStep(text, keyword, currentContext)
			test 'Example 3: 14, 0, 14' do(suite)
				text = 'I add <a> and <b>'
				text = text.replaceAll('<a>', 14)
				text = text.replaceAll('<b>', 0)
				text = text.replaceAll('<result>', 14)
				keyword = 'Given'
				currentContext = contexts.c2.context
				await executeStep(text, keyword, currentContext)
				text = 'I should obtain <result>'
				text = text.replaceAll('<a>', 14)
				text = text.replaceAll('<b>', 0)
				text = text.replaceAll('<result>', 14)
				keyword = 'Then'
				currentContext = contexts.c2.context
				await executeStep(text, keyword, currentContext)
			test 'Example 4: 15, -1, 14' do(suite)
				text = 'I add <a> and <b>'
				text = text.replaceAll('<a>', 15)
				text = text.replaceAll('<b>', -1)
				text = text.replaceAll('<result>', 14)
				keyword = 'Given'
				currentContext = contexts.c2.context
				await executeStep(text, keyword, currentContext)
				text = 'I should obtain <result>'
				text = text.replaceAll('<a>', 15)
				text = text.replaceAll('<b>', -1)
				text = text.replaceAll('<result>', 14)
				keyword = 'Then'
				currentContext = contexts.c2.context
				await executeStep(text, keyword, currentContext)
			test 'Example 5: 16, -2, 14' do(suite)
				text = 'I add <a> and <b>'
				text = text.replaceAll('<a>', 16)
				text = text.replaceAll('<b>', -2)
				text = text.replaceAll('<result>', 14)
				keyword = 'Given'
				currentContext = contexts.c2.context
				await executeStep(text, keyword, currentContext)
				text = 'I should obtain <result>'
				text = text.replaceAll('<a>', 16)
				text = text.replaceAll('<b>', -2)
				text = text.replaceAll('<result>', 14)
				keyword = 'Then'
				currentContext = contexts.c2.context
				await executeStep(text, keyword, currentContext)
			test 'Example 6: 17, -3, 14' do(suite)
				text = 'I add <a> and <b>'
				text = text.replaceAll('<a>', 17)
				text = text.replaceAll('<b>', -3)
				text = text.replaceAll('<result>', 14)
				keyword = 'Given'
				currentContext = contexts.c2.context
				await executeStep(text, keyword, currentContext)
				text = 'I should obtain <result>'
				text = text.replaceAll('<a>', 17)
				text = text.replaceAll('<b>', -3)
				text = text.replaceAll('<result>', 14)
				keyword = 'Then'
				currentContext = contexts.c2.context
				await executeStep(text, keyword, currentContext)
			test 'Example 7: 18, -4, 14' do(suite)
				text = 'I add <a> and <b>'
				text = text.replaceAll('<a>', 18)
				text = text.replaceAll('<b>', -4)
				text = text.replaceAll('<result>', 14)
				keyword = 'Given'
				currentContext = contexts.c2.context
				await executeStep(text, keyword, currentContext)
				text = 'I should obtain <result>'
				text = text.replaceAll('<a>', 18)
				text = text.replaceAll('<b>', -4)
				text = text.replaceAll('<result>', 14)
				keyword = 'Then'
				currentContext = contexts.c2.context
				await executeStep(text, keyword, currentContext)
	describe 'Scenario Outline: Adding two numbers works on many cases' do
		describe 'Examples group 0: Basic examples for concat' do
			test 'Example 0: al, pha, alpha' do(suite)
				text = 'I concatenate <a> and <b>'
				text = text.replaceAll('<a>', '\\"al\\"')
				text = text.replaceAll('<b>', '\\"pha\\"')
				text = text.replaceAll('<result>', '\\"alpha\\"')
				keyword = 'Given'
				currentContext = contexts.c3.context
				await executeStep(text, keyword, currentContext)
				text = 'The \\"str\\" in context is <result>'
				text = text.replaceAll('<a>', '\\"al\\"')
				text = text.replaceAll('<b>', '\\"pha\\"')
				text = text.replaceAll('<result>', '\\"alpha\\"')
				keyword = 'Then'
				currentContext = contexts.c3.context
				await executeStep(text, keyword, currentContext)
			test 'Example 1: ab, dellah, abdellah' do(suite)
				text = 'I concatenate <a> and <b>'
				text = text.replaceAll('<a>', '\\"ab\\"')
				text = text.replaceAll('<b>', '\\"dellah\\"')
				text = text.replaceAll('<result>', '\\"abdellah\\"')
				keyword = 'Given'
				currentContext = contexts.c3.context
				await executeStep(text, keyword, currentContext)
				text = 'The \\"str\\" in context is <result>'
				text = text.replaceAll('<a>', '\\"ab\\"')
				text = text.replaceAll('<b>', '\\"dellah\\"')
				text = text.replaceAll('<result>', '\\"abdellah\\"')
				keyword = 'Then'
				currentContext = contexts.c3.context
				await executeStep(text, keyword, currentContext)
"
`;

exports[`Cucumber plugin  > support other languages 1`] = `
"import {parseFeatureIntoSteps} from \\"imba/plugin\\"

const modules = import.meta.glob('./step_definitions/**/*.imba', {eager: yes})
const baseContexts = import.meta.glob('./**/*.context.imba', {eager: yes})
let s
let c
let elName
let contexts = {}
let text
let stepRes
let keyword
let currentContext

const decorators = {}


def executeStep(text, keyword, context)
	let s = global.Steps.find(text, keyword).stepDefinition
	if const ext = s.target.extendContext
		await ext(context)
	let res = await s.target[s.fname]
		.apply(
			context,
			# variables
			[s.cucumberExpression.match(text).map(do $1.getValue!)]
		)

	# Assertions throw an error if accessed (they use a proxy)
	return if res..__flags
	return unless const features = res..__features
	await features.reduce(&, do 1) do(acc, feature)
		await acc
		const steps = await parseFeatureIntoSteps(feature)
		await steps.reduce(&, do 1) do(acc, step)
			await acc
			await executeStep(step.text, step.keyword, context)


beforeAll do(suite)
beforeEach do(suite)

	let contextName = 'BaseContext'
	const baseContext = Object.values(baseContexts).find(do $1[contextName])..[contextName]
	elName = 'مثال: ااا'
	if suite.meta.name == elName or suite.meta.suite..suite..name == elName
		suite.meta.decorators = decorators
		contexts.c0 = {
			context: global.__ccContext(baseContext, suite.meta)
			name: elName
		}


describe 'Feature: برنامجي' do
	test 'مثال: ااا' do
		text = 'أبدأ البرنامج'
		keyword = 'عندما'
		currentContext = contexts.c0.context
		await executeStep(text, keyword, currentContext)
"
`;
